nemo/core/neural_types/nmtensor_registry.py

```diff
@@ -0,0 +1,88 @@
+# =============================================================================
+# Copyright (c) 2020 NVIDIA. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+
+class NmTensorNameRegistry:
+    def __init__(self):
+        """
+            Constructor. Initializes the NmTensorNameRegistry. Reserves the default 'loss' name.
+
+            TODO: We should be recording the tensors of each graph rather than all the tensors.
+        """
+        # Create the nmtensor_naming_dict
+        # which contains a mapping of str to NMTensor.unique_name
+        self._nmtensor_naming_dict = {"loss": "loss"}  # Reserve keyname of 'loss'
+        self._nmtensor_uniname_dict = {"loss": None}
+
+    @property
+    def unique_names(self):
+        """Returns the set of all NmTensors.unique_names + 'loss'
+        """
+        return self._nmtensor_uniname_dict.keys()
+
+    def register(self, tensor: 'NmTensor'):
+        """Helper function to register a newly created NmTensor by adding it to self.__nmtensor_uniname_dict.
+        Should be called from NmTensor.__init__()
+
+        args:
+            tensor (NmTensor): The tensor to be registered.
+        """
+
+        # Check if object is already in a set.
+        if tensor.unique_name in self._nmtensor_uniname_dict:
+            pass
+
+        # Finally, add object to the set.
+        self._nmtensor_uniname_dict[tensor.unique_name] = tensor
```

Comment:
This looks like a "strong reference" to tensor object...

Reply:
Removing this reference and switching _nmtensor_uniname_dict to be a set()